### ISSUE TITLE
Issue when resetting certain numbers with a default value

### ISSUE THREAD

**Issue by ayush-shta** [rjsf-team/react-jsonschema-form#4202]
### Prerequisites

- [X] I have searched the existing issues
- [X] I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [X] I have read the [documentation](https://rjsf-team.github.io/react-jsonschema-form/docs)
- [X] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net), [Codesandbox.io](https://codesandbox.io) or preferably a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### What theme are you using?

core

### Version

5.x

### Current Behavior

Numbers not resetting when the number entered in that field ends with the default value provided for that field in the schema

### Expected Behavior

Values should reset as expected and should not depend upon what the default value is

### Steps To Reproduce

I tried to replicate this issue in the playground. [Link](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJudW1iZXIiOjF9LCJzY2hlbWEiOnsidHlwZSI6Im9iamVjdCIsInRpdGxlIjoiTnVtYmVyIGZpZWxkcyAmIHdpZGdldHMiLCJwcm9wZXJ0aWVzIjp7Im51bWJlciI6eyJ0aXRsZSI6Ik51bWJlciIsInR5cGUiOiJudW1iZXIiLCJkZWZhdWx0IjoxfX19LCJ1aVNjaGVtYSI6eyJpbnRlZ2VyIjp7InVpOndpZGdldCI6InVwZG93biJ9LCJudW1iZXJFbnVtUmFkaW8iOnsidWk6d2lkZ2V0IjoicmFkaW8iLCJ1aTpvcHRpb25zIjp7ImlubGluZSI6dHJ1ZX19LCJpbnRlZ2VyUmFuZ2UiOnsidWk6d2lkZ2V0IjoicmFuZ2UifSwiaW50ZWdlclJhbmdlU3RlcHMiOnsidWk6d2lkZ2V0IjoicmFuZ2UifX0sInRoZW1lIjoiZGVmYXVsdCIsImxpdmVTZXR0aW5ncyI6eyJzaG93RXJyb3JMaXN0IjoidG9wIiwiZXhwZXJpbWVudGFsX2RlZmF1bHRGb3JtU3RhdGVCZWhhdmlvciI6eyJhcnJheU1pbkl0ZW1zIjp7InBvcHVsYXRlIjoicG9wdWxhdGUiLCJtZXJnZUV4dHJhRGVmYXVsdHMiOmZhbHNlfSwiYWxsT2YiOiJza2lwRGVmYXVsdHMiLCJlbXB0eU9iamVjdEZpZWxkcyI6InBvcHVsYXRlQWxsRGVmYXVsdHMifX19)

1. Lets say I have a schema including one of the property to be like the one below. Here the default value of this field is 1.
```
  "number": {
      "title": "Number",
      "type": "number",
      "default": 1
    },
```

2. Now I type in the value for this filed as any number ending with the default value (for example, 11, 231, 0.1, 1.1, 2.1, 0.0001)
3. After that, lets click the button to reset the form
4. The value for that field does not reset
5. However if we type in any other number that does not end with the default value it will reset as expected

### Environment

```markdown
- OS:
- Node:
- npm:
```

**Comment by nickgros** [comment]
Looks like a bug in the regular expression check in NumberField: https://github.com/rjsf-team/react-jsonschema-form/blob/ec932db942dd046640303056c89e3501b16ec469/packages/core/src/components/fields/NumberField.tsx#L74

**Comment by nickgros** [comment]
Proposed fix here: https://github.com/rjsf-team/react-jsonschema-form/pull/4220

**Comment by heath-freenome** [comment]
Fixed by #4420
